src/engraving/io/mscreader.cpp

```diff
@@ -0,0 +1,379 @@
+/*
+ * SPDX-License-Identifier: GPL-3.0-only
+ * MuseScore-CLA-applies
+ *
+ * MuseScore
+ * Music Composition & Notation
+ *
+ * Copyright (C) 2021 MuseScore BVBA and others
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+#include "mscreader.h"
+
+#include <QXmlStreamReader>
+#include <QFile>
+#include <QFileInfo>
+#include <QDir>
+#include <QDirIterator>
+
+#include "thirdparty/qzip/qzipreader_p.h"
+
+#include "log.h"
+
+//! NOTE The current implementation resolves files by extension.
+//! This will probably be changed in the future.
+
+using namespace mu::engraving;
+
+MscReader::MscReader(const Params& params)
+    : m_params(params)
+{
+}
+
+MscReader::~MscReader()
+{
+    close();
+}
+
+void MscReader::setParams(const Params& params)
+{
+    IF_ASSERT_FAILED(!isOpened()) {
+        return;
+    }
+
+    if (m_reader) {

```

Comment:
why would we need to delete m_reader here?

If it's really necessary, maybe it makes sense to call "close()" here?

Reply:
I think, call `close` is bad idea, and should be already closed at the moment. 
But, we should delete reader, because it may be created, but mode may be changed, so may be needs other reader implementation.